packages/terser/src/worker-pool.ts

```diff
@@ -0,0 +1,117 @@
+import { Worker } from 'worker_threads';
+import { cpus } from 'os';
+import { EventEmitter } from 'events';
+
+import serializeJavascript from 'serialize-javascript';
+
+import {
+  WorkerCallback,
+  WorkerContext,
+  WorkerOutput,
+  WorkerPoolOptions,
+  WorkerPoolTask
+} from './type';
+
+const symbol = Symbol.for('FreeWoker');
+
+export class WorkerPool extends EventEmitter {
+  protected maxInstances: number;
+
+  protected filePath: string;
+
+  protected tasks: WorkerPoolTask[] = [];
+
+  protected workers = 0;
+
+  constructor(options: WorkerPoolOptions) {
+    super();
+
+    this.maxInstances = options.maxWorkers || cpus().length;
+    this.filePath = options.filePath;
+
+    this.on(symbol, () => {
+      if (this.tasks.length > 0) {
+        this.run();
+      }
+    });
+  }
+
+  add(context: WorkerContext, cb: WorkerCallback) {
+    this.tasks.push({
+      context,
+      cb
+    });
+
+    if (this.workers >= this.maxInstances) {
+      return;
+    }
+
+    this.run();
+  }
+
+  async addAsync(context: WorkerContext): Promise<WorkerOutput> {
+    return new Promise((resolve, reject) => {
+      this.add(context, (err, output) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        if (!output) {
+          reject(new Error('The output is empty'));
+          return;
+        }
+
+        resolve(output);
+      });
+    });
+  }
+
+  private run() {
+    if (this.tasks.length === 0) {
+      return;
+    }
+
+    const task = this.tasks.shift();
+
+    if (typeof task === 'undefined') {
+      return;
+    }
+
+    this.workers += 1;
+
+    let called = false;
+    const callCallback = (err: Error | null, output?: WorkerOutput) => {
+      if (called) {
+        return;
+      }
+      called = true;
+
+      this.workers -= 1;
+
+      task.cb(err, output);
+      this.emit(symbol);
+    };
+
+    const worker = new Worker(this.filePath, {
```

Comment:
Are Workers ever reused and returned to the pool? If not, this would be better to avoid the overhead of creating new workers.

Reply:
No, it is more or less a fake worker pool, to ensure that not more than x workers run i parallel and the event loop is not blocked, for other applications or tasks, by blocking the event loop, with open workers.

https://nodejs.org/en/docs/guides/dont-block-the-event-loop/

Reply:
Ah OK, maybe a future enhancement would be to create a true worker pool like the one shown here: https://nodejs.org/api/async_context.html#using-asyncresource-for-a-worker-thread-pool